### ISSUE TITLE
Phase 5: Replace VitTracker with LiteTrack-B4 or HiT

### ISSUE THREAD

**Issue by tuky191** [tuky191/haptic-track#128]
## Context

VitTracker (0.7MB, OpenCV DNN, CPU) is the weakest model in accuracy terms. Multiple edge-optimized trackers now offer better tracking with acceptable size/latency.

**Important caveat**: the +33% tracking accuracy claim is from Jetson benchmarks on LaSOT/GOT-10k, not Adreno 740 on handheld footage. The actual VT pain points are box growth and small-object failure — a better AUC may or may not fix those. Gate on offline measurement against our scenarios.

## Current state

- **VitTracker** (0.7MB ONNX, ~5ms CPU, 48.6 AUC on LaSOT)
- Integrated via OpenCV `TrackerVit` wrapper
- Provides confidence score (critical for drift detection)
- Known issues: regression head lets boxes grow (mitigated by #120 reseat), dies fast on small/transparent objects (#59)

## Candidates

| Tracker | Size | Edge FPS | Accuracy | Export | Confidence? |
|---|---|---|---|---|---|
| VitTracker (current) | 0.7MB | ~200 CPU | 48.6 AUC | ONNX/OpenCV | Yes |
| **LiteTrack-B4** | ~5MB | 44 Jetson Orin NX | 65.2 AO GOT-10k | ONNX | Needs verification |
| **HiT / DyHiT** | ~5-8MB | 61-111 Jetson AGX | 62-65% AUC LaSOT | Not documented | Needs verification |
| **MixFormerV2-S** | ~4MB | 45 CPU | 60.6% AUC | ONNX | Yes (distillation) |
| **UETrack-T** | ~6MB | 77 Jetson AGX | competitive | Code only (Mar 2026) | Unknown |
| **MVT** | ~5MB | 70 CPU | competitive | ONNX + TRT | Uses MobileViT backbone |

## Before committing to a swap

1. **Check OpenCV DNN compatibility**: can the candidate ONNX models be loaded via OpenCV DNN (which we already have)? Strongly prefer OpenCV-DNN-compatible options — avoids adding ONNX Runtime as a new dependency.
2. **Run candidate against our scenarios offline**: chair_living_room, mouse_desk_rotation, bottle captures. Measure VT loss rate, box-growth events, and confidence-score quality. The real metric is "fewer scenarios timing out due to VT failure," not LaSOT AUC.
3. **Verify confidence score**: the replacement must provide a reliable confidence/quality signal. NanoTrack was ruled out because it always returns 0.9 (no loss detection possible).

## Key risks

- **Benchmark numbers don't transfer**: Jetson Orin NX has ~25 TOPS INT8 vs Adreno 740 ~15-20 TOPS. Latency on our SoC will be slower than the candidate table implies.
- **OpenCV integration loss**: If candidates don't load via OpenCV DNN, we need ONNX Runtime for Android — a new GPU/NNAPI integration path, new failure modes, and a new APK artifact.
- **Latency budget**: VitTracker runs every frame at ~5ms. Replacement must stay under ~15ms.
- **Different failure modes**: New tracker may have its own box-stability or drift patterns. VT reseat logic (#120) and aggressive drift detection (#45) may need adaptation — but if the new tracker is better, some of that complexity can be removed (a real win).

## Scope

- Evaluate candidates for OpenCV DNN compatibility (spike first)
- Replace `VisualTracker.kt` model loading and inference
- Adapt confidence thresholds (`MIN_CONFIDENCE`, `VT_SKIP_STABLE_CONFIDENCE`)
- Re-evaluate drift detection logic — a better tracker may need less aggressive checking
- Re-evaluate VT reseat logic — may be unnecessary if new tracker doesn't grow boxes

## Validation

- All scenario replay tests pass
- On-device: tracking stability on all test videos (compare VT loss rate before/after)
- Frame timing: must stay under 15ms per tracker update
- Specifically test: chair_living_room (box growth), mouse_desk_rotation (small-object failure)
- Drift detection still works (template self-verification + detector cross-check)

## Gated on

- Phases 0/1/4 shipped first (lower risk, higher confidence)
- Offline benchmark infrastructure built (also needed for Phase 3)

## Saves

- No APK savings (0.7MB → ~5MB, actually grows)
- **Accuracy savings**: fewer VT losses → fewer re-acquisition cycles → smoother user experience
- Potentially simplifies drift detection + VT reseat logic (code complexity reduction)

**Comment by tuky191** [comment]
## Assessment

Net positive but the "+33% tracking accuracy" claim doesn't transfer cleanly from the candidate table to our actual scenarios.

### The numbers are from different hardware and different benchmarks

- LiteTrack-B4: 44 FPS on Jetson Orin NX, 65.2 AO on GOT-10k
- HiT/DyHiT: 61-111 FPS on Jetson AGX, 62-65% AUC on LaSOT
- Our environment: Adreno 740 on Snapdragon 8 Gen 2, handheld phone footage

Jetson Orin NX has ~25 TOPS INT8; Adreno 740 has ~15-20 TOPS depending on workload. Latency on our SoC will be slower than the table implies. And LaSOT/GOT-10k are not handheld-phone-tracking-a-locked-subject benchmarks — they're general short-term tracking with curated targets.

**Recommendation:** before committing to a tracker, run the candidate against our captured scenarios offline and measure the actual VT loss rate, box-growth rate, and confidence-score quality. If the gain on our scenes isn't visibly better than VitTracker post-#120 reseat, the migration cost outweighs the benefit.

### The actual VT pain isn't "low AUC" — it's box growth and small-object failure

What's visibly broken about VitTracker in our usage:
1. Regression head lets boxes grow on stable subjects (mitigated, not eliminated, by #120 reseat)
2. Confidence drops below 0.25 within 2-3s on small/transparent objects (bottles, glasses) (#59)

LaSOT AUC measures short-term tracking accuracy on a different distribution. A tracker with +33% AUC may or may not fix issues 1 and 2. **Concrete validation question:** does LiteTrack-B4 (or HiT) produce a stable bounding box on the chair_living_room and mouse_desk_rotation captures, where VitTracker fails today? If yes, ship it. If no, we're trading one set of failures for another.

### ONNX Runtime for Android is a meaningful new dependency

The issue mentions "raw ONNX Runtime for Android, which is a different runtime from TFLite" as if it's a minor implementation detail. It's not — it's a new GPU/NNAPI integration path, new failure modes (model loading, delegate fallback, op support), and a new artifact in the APK. **Worth investigating first:** can the candidate ONNX models be loaded via OpenCV DNN (which we already have for VitTracker)? If yes, the integration cost drops to "swap the .onnx file." If no, the ONNX Runtime dependency is real overhead and that should be in the trade-off.

### Recommendation

Before committing to a tracker swap:
1. Verify which candidates load via OpenCV DNN vs requiring ONNX Runtime. Strongly prefer OpenCV-DNN-compatible options.
2. Run the candidate against the chair, mouse, and bottle scenarios offline — measure VT loss rate and box-growth events specifically. The real metric isn't AUC, it's "fewer scenarios timing out due to VT failure."
3. Re-evaluate whether VT reseat (#120) and aggressive drift detection (#45) can be relaxed with a better tracker. That's a real win — simplification, not just accuracy.

The phase is worth doing, but only after Phase 0/1/4 have shipped and we have offline-benchmark infrastructure built (which is also useful for Phase 3).

**Comment by tuky191** [comment]
Updated: clarified that the gate is "Phases 0/1/4 shipped first" — Phase 4b (#130, multi-task training project) is explicitly NOT a gate for this phase.

**Comment by tuky191** [comment]
## Implementation status (2026-05-01)

Branch: `feat/litetrack-b4-128`

### What's done

**Off-device feasibility** ✅
- Pre-converted TFLite models found in PINTO0309 model zoo (LiteTrack #373)
- Two-model architecture verified:
  - **Backbone** (template encoder): `[1,127,127,3]` → `[1,8,8,96]` — run once at lock
  - **Track** (search decoder): `[1,8,8,96]` + `[1,256,256,3]` → cls `[1,16,16,1]` + reg `[1,16,16,4]` — run every frame
- Output format confirmed from ONNX graph analysis:
  - cls: raw logits × 0.1 temperature scaling → apply sigmoid for confidence
  - reg: FCOS-style LTRB distances in search-image pixels (relu'd × 16 stride)
- FP16 TFLite models: backbone 2.5MB + track 4.0MB = **6.5MB total**

**Android integration** ✅
- `VisualTracker.kt` rewritten: OpenCV `TrackerVit` → two TFLite GPU interpreters
- Old `object_tracking_vittrack_2023sep.onnx` (0.7MB) removed
- Asset files: `litetrack_b4_backbone_fp16.tflite`, `litetrack_b4_track_fp16.tflite`
- SiamFC-style cropping: context-padded template crop + proportional search region
- Hann window applied to score map for temporal stability
- Init-time NaN check catches GPU delegate failures early
- Same API surface (`init`, `update`, `stop`, `isActive`, `TrackerResult`)
- Build passes, all 301 unit tests pass
- Net APK delta: +5.8MB

### What's needed: on-device validation

1. **GPU delegate loading** — check logcat for `TFLiteGPU` tag:
   ```
   adb logcat -s "TFLiteGPU" -s "VisualTracker"
   ```
   Must show `GPU delegate active for LiteTrack-Backbone` and `LiteTrack-Track`. If it falls back to CPU, note latency.

2. **NaN check** — init logs `Backbone output NaN` if GPU miscompiles (same class as OSNet-IBN #117, MobileCLIP #126). If NaN, try CPU fallback.

3. **Basic tracking** — tap to lock, verify box follows target. Watch for:
   - Box stability (does it grow like VitTracker?)
   - Confidence scores (are they meaningful, or flat like TrackerNano was?)
   - Drift behavior (does template mismatch fire appropriately?)

4. **VideoReplayTest suite** — quantitative comparison:
   ```
   adb shell am instrument -w -e class com.haptictrack.VideoReplayTest com.haptictrack.test/androidx.test.runner.AndroidJUnitRunner
   ```
   Compare tracking rates against VitTracker baseline.

5. **Latency** — VitTracker was ~5ms. LiteTrack-B4 has more parameters. Must stay under ~15ms.

### LTRB decoding reference (for debugging)

```
cell_center_x = (grid_x + 0.5) * 16
cell_center_y = (grid_y + 0.5) * 16
bbox_left  = cell_center_x - reg[0]
bbox_top   = cell_center_y - reg[1]  
bbox_right = cell_center_x + reg[2]
bbox_bottom = cell_center_y + reg[3]
```
Then map from 256×256 search coords back to image coords using the crop scale factor.

### Risk: reg channel order

The LTRB channel mapping `[left, top, right, bottom]` was inferred from ONNX graph analysis (FCOS-style head with relu + ×16 stride). If tracking looks wrong on device (box offset in one direction, or wrong aspect ratio), the channel order might be `[top, left, bottom, right]` or similar — swap channels and retry.

**Comment by tuky191** [comment]
Closed — VitTracker stays. Benchmarked HiT-Small (TFLite/ORT/NCNN) and LightTrack. Transformer trackers can't get mobile GPU acceleration (MatMul/attention ops unsupported by TFLite GPU, NNAPI, NCNN Vulkan). Best HiT result: 28ms ORT CPU vs VitTracker's 5ms. LightTrack (PINTO, mislabeled as LiteTrack-B4) kept on feat/litetrack-b4-128 branch as backup.
